### ISSUE TITLE
Option to Add next steps... unavailable for submodel

### ISSUE THREAD

**Issue by jacovandermolen** [trevorsandy/lpub3d#632]
### Subject
Option to Add next steps... unavailable for submodel
(and multistep commands messed up)

### Environment
* Version of LPub3D - 2.4.5.24.3232 (and some versions before that)
* Operating system - Windows 10

### Steps to reproduce
Upon encountering a page where a submodel starts with multiple steps, there used to be an option to "Add next steps..." with a right-mouse click to pop up the menua and then you could choose how many steps you wanted to add to multistep page.

### Expected behaviour
The Add next steps... option should be available on right-mouse click

### Actual behaviour
Choosign Add next step once works, but Add next step again and the multistep meta commands are messed up
![image](https://user-images.githubusercontent.com/32261647/205052060-5ef4352a-9bc8-437e-8ae6-817e48d9d6fe.png)

### Workaround
Add one next step and manually move the Multistep end command where you want it

### Solution suggestion
Please fix option to add multiple next steps and fix the placement Multistep metacommands

**Comment by trevorsandy** [comment]
Thank you for reporting this behaviour. 

> Option to Add next steps... unavailable for submodel.

Are you sure you are not at the end of the submodel ?

Here are three scenarios that affect the context menu actions

1. First page of submodel - add previous step action is not displayed
![Screenshot - 01_12_2022 , 14_54_48](https://user-images.githubusercontent.com/13388970/205071603-33639e7f-796b-4e56-9e4e-a281af240cfe.png)

2. Middle page of submodel - add previous, next step and next steps (when there is more than 1) displayed
![Screenshot - 01_12_2022 , 14_55_37](https://user-images.githubusercontent.com/13388970/205072074-b386ec58-72b6-4086-8eec-bee3054ea732.png)

3. Last page (with parts) of submodel - add next steps not displayed because there is less than 2 next steps in the submodel
![Screenshot - 01_12_2022 , 14_53_33](https://user-images.githubusercontent.com/13388970/205072352-2e8ce96e-fd90-4d44-bacf-7d1c94a9f7c8.png)

> Choosign Add next step once works, but Add next step again and the multistep meta commands are messed up

I'll take a look.

Cheers,

**Comment by trevorsandy** [comment]
> Choosing Add next step once works, but Add next step again and the multistep meta commands are messed up

This behaviour has been corrected in #637 

Cheers,

**Comment by jacovandermolen** [comment]
OK, DevOp version 2.4.5.42.3250 corrected this in most of my models.
However a new buildmod error is encounterd.
As are some new problems in another model, that has not yet have any LPub commands
![image](https://user-images.githubusercontent.com/32261647/205446756-a866e06e-caeb-4008-a0f5-a97ef9d1f98c.png)
![image](https://user-images.githubusercontent.com/32261647/205446765-bf91b611-2ea0-46cd-85d1-42035cc7111d.png)
![image](https://user-images.githubusercontent.com/32261647/205446813-304d4d81-d761-4f7b-85ad-42c10549200e.png)

**Comment by jacovandermolen** [comment]
> > Choosign Add next step once works, but Add next step again and the multistep meta commands are messed up

This is fixed.
Still no Add next steps... in a submodel with more than 2 steps.
See my model file (e-mail)

**Comment by trevorsandy** [comment]
> Still no Add next steps... in a submodel with more than 2 steps.

Very well, I'll take another look.

Cheers,

**Comment by trevorsandy** [comment]
I'm still not able to reproduce this behaviour. 

The model file you are using in the emailed error image has been modified from the version I have. In my version, there is a build modification in the first step of `submodel-36.ldr`. I've recreated the model file scenario using the image and the model file I have. 

My result is as expected.

![Screenshot - 04_12_2022 , 11_03_44](https://user-images.githubusercontent.com/13388970/205484933-d8024a3e-54f8-4199-8d13-6d0600195b2a.png)

Here is my model file that creates the scenario in the error image you sent - I'm using your header.ldr.

```
0 Name: stabilization.ldr
0 Author: LPub3D [lp3d]
0 !LICENSE Redistributable under CCAL version 2.0 : see CAreadme.txt
0 !LPUB SUBMODEL_DISPLAY MODEL_SCALE GLOBAL  0.9000
0 !LPUB SUBMODEL_DISPLAY SHOW GLOBAL TRUE
0 !LPUB BUILD_MOD_ENABLED GLOBAL TRUE
0 !LPUB CONTINUOUS_STEP_NUMBERS GLOBAL TRUE
0 !LPUB PARSE_NOSTEP GLOBAL TRUE
0 !LPUB INCLUDE "header.ldr"
0 !LPUB INSERT PAGE
0 STEP
1 4 20 -40 -80 0 -1 0 1 0 0 0 0 1 32526.dat
0 STEP
1 2 -12 -8 0 0 -1 0 -1 0 0 0 0 -1 stabilization-sub.ldr
0 STEP
1 2 0 0 0 0 -1 0 -1 0 0 0 0 -1 32526.dat
0 NOFILE
0 FILE stabilization-sub.ldr
0 Name: stabilization-sub.ldr
0 Author: LPub3D [lp3d]
0 !LICENSE Redistributable under CCAL version 2.0 : see CAreadme.txt
0 ROTATION CENTER 0 0 0 1 "Custom"
0 ROTATION CONFIG 0 0
1 2 0 0 0 0 -1 0 -1 0 0 0 0 -1 32526.dat
0 ROTSTEP 180 180 0 REL
1 0 10 -40 -80 -1 0 0 0 1 0 0 0 -1 2780.dat
0 STEP
1 4 20 -40 -80 0 -1 0 1 0 0 0 0 1 32526.dat
0 ROTSTEP END
1 71 20 0 0 -1 0 0 0 0 1 0 1 0 32054.dat
0 STEP
1 14 40 -40 -60 0 -1 0 1 0 0 0 0 1 18654.dat
0 STEP
1 7 40 -40 -60 -1 0 0 0 0 -1 0 -1 0 32054.dat
0 NOFILE
```

Let me know if you are getting the same result as me using my file. If yes, then send along a version of your model file that produces the unexpected behaviour.

Cheers,

**Comment by jacovandermolen** [comment]
OK, Add next steps... does work indeed with your test model.
The only thing I get now is
[image: image.png]

And

[image: image.png]

header.ldr is in the same folder and does work either way.

I will check all my files and send you one e-mail regarding the errors that
remain.
We are getting there though! :-)

Op zo 4 dec. 2022 om 11:13 schreef Trevor SANDY ***@***.***>:

**Comment by trevorsandy** [comment]
> The only thing I get now is

That's my mistake. I forgot to add the fix for .ldr files, I only added it for mpd files. Just remove the quotes(") from header.ldr and you'll be good to go.

> We are getting there though! :-)

Indeed - I'm looking at the unformatting crash behaviour ATM. It's strange but I can't reproduce an AbnormalEnd but I did find a conflict with the build modification setting.

Cheers,

**Comment by jacovandermolen** [comment]
> Just remove the quotes(") from header.ldr and you'll be good to go.

Not yet fixed in 2.4.5.49.3257?

Concerning some other errors, 2.4.5.49.3257 did fix build mods in one of my models I have tested just now.
I will test my other models!

Looking good!

**Comment by trevorsandy** [comment]
> > Just remove the quotes(") from header.ldr and you'll be good to go.

> Not yet fixed in 2.4.5.49.3257?

Yes - this is fixed in the latest DevOps build. The _failed to resolve file at path_ message should not be triggered if you have quotes (") around your INCLUDE file name in an LDR file. 

My quote above was a result of #631 (first commit) not being applied to LDR files. In the first fix, my regular expression used to grab the INCLUDE file also grabbed the quotes so the value passed to the file path resolution call ended up being enclosed by a pair of double quotes ("") hence, the error when quotes are used. I've updated the RegExp to ignore double(") or single (') quotes in the latest DevOps build. So both LDR and MPD files with a quoted INCLUDE file should load without error.

> Concerning some other errors, 2.4.5.49.3257 did fix build mods in one of my models I have tested just now.

Excellent. 

These are my outstanding items for the next official release

1. Generated _BUILD_MOD REMOVE/APPLY action command must be placed before step parts.._ message triggered by a build modification in a called out submodel which is itself a submodel an also in a build modification. This is the last of the 3 items you discussed in #625
2. Called out build modifications when jumping forward where the steps skipped with the forward jump were not previously generated normally (i.e. incrementally).
3. The LPub3D plug logo is not generated in some cases
4. Build modification menu actions in the Visual Editor are enabled when they should not be.
5. Visual Editor does not display applied build modifications in multi-step pages when the build mod step is not the last step on the multi-step page.

Cheers,
